sympy/polys/matrices/normalforms.py

```diff
@@ -336,7 +339,97 @@ def add_columns_mod_R(m, R, i, j, a, b, c, d):
     return DomainMatrix(W, (m, m), ZZ).to_dense()
 
 
-def hermite_normal_form(A, *, D=None, check_rank=False):
+class HNFMode(Enum):
+    r"""
```

Comment:
What about the `Enum`? It's possible that with this I was thinking more like a software engineer, and less like a user.

My impression is that SymPy doesn't use a lot of `Enum`s. Typically a function or method that needs a mode switch seems to take a string. E.g. `limit` accepts `dir="+"` or `dir="-"`. `DomainMatrix` accepts `fmt="sparse"` or `fmt="dense"`.

From the user's perspective, it would certainly be handier to be able to just write `mode="a"` than to have to (a) import an enum class and then (b) write the much longer `mode=HNFMode.ABOVE`. In retrospect I'm thinking I might like the string approach better. Any opinions?

Reply:
I don't know anything about Enum but I am supportive of the text mode. Possibilities include "NEWS" to indicate a map direction (N=north/above pivot); or glyph-like "^v<>" to point where the reduced entries should go.

Reply:
Can we set the Enum values to strings? That could be a middle ground. I think typing a string is more convenient as a user, but the Enum can also be useful in that it makes it easy to see what the options are from tab completion. 

Allowing different equivalent variants of the same string is useful too (like `"ABOVE"` or `"above"`, or @smichr's suggestions). I don't know if Enum supports that, but it could accept it in the function.

Reply:
I like the compass and glyph-like alternative strings. A user could easily forget whether it was supposed to be "above" or "up", but "^" is clear.

Reply:
> Can we set the Enum values to strings?

I'm new to `Enum`s myself, and don't know if there's a handy way to check them simultaneously as a string and as an actual `Enum` member. You can set strings as values, but the enum members still do not test equal to strings:

```python
>>> from enum import Enum
>>> class Foo(Enum):
...     BAR = "BAR"
...     SPAM = "SPAM"
>>> "BAR" == Foo.BAR
False
```

If you want a single test that accepts both strings and enum members, maybe an `or` is the best you can do?

```python
>>> test = lambda x: x is Foo.BAR or x == Foo.BAR.value
>>> test("BAR")
True
>>> test(Foo.BAR)
True
```